core/src/main/kotlin/fr/sncf/osrd/stdcm/graph/VisitedNodes.kt

```diff
@@ -28,86 +25,151 @@ import kotlin.Double.Companion.POSITIVE_INFINITY
+        // the time t is visited at f(t) seconds of total running time duration
+        private val visitedWithTravelTime: LinearFunction,
+    ) : ConditionallyVisitedRange {
+        override fun isVisited(
+            range: Range<Double>,
+            parameters: Parameters,
+        ): Boolean {
+            val runningTimeForVisited = visitedWithTravelTime.apply(range.lowerEndpoint())
+            return runningTimeForVisited <= parameters.nodeCost
+        }
+
+        override fun mergeWith(other: ConditionallyVisitedRange?): ConditionallyVisitedRange {
+            if (other is VisitedWithAddedTravelTime) {
+                return if (visitedWithTravelTime < other.visitedWithTravelTime) this else other
+            }
+            return other ?: this
+        }
+    }
+
+    private val visitedRangesPerLocation =
+        mutableMapOf<Fingerprint, RangeMap<Double, ConditionallyVisitedRange>>()
 
     /** Returns true if the input has already been visited */
     fun isVisited(
         parameters: Parameters,
     ): Boolean {
-        val alreadyVisitedTimes =
-            visitedRangesPerLocation[parameters.fingerprint!!] ?: ImmutableRangeSet.of()
-        val alreadyReachableRanges =
-            reachableRangesPerLocation[parameters.fingerprint!!]
-                ?: ImmutableRangeMap.of(Range.all<Double>(), LinearFunction(POSITIVE_INFINITY))
+        val visitedRanges = visitedRangesPerLocation[parameters.fingerprint!!] ?: return false
+        val timeData = parameters.timeData
 
         val visitingRange =
-            Range.closed(parameters.startTime, parameters.startTime + parameters.duration)
-
-        // List the ranges that are being explored, outside the previously visited ranges
-        val rangesToCheck = TreeRangeSet.create<Double>()
-        rangesToCheck.add(visitingRange)
-        rangesToCheck.removeAll(alreadyVisitedTimes)
-
-        // then check what's left against the reachable ranges
-        for (range in rangesToCheck.asRanges()) {
-            val alreadyReachableMap = alreadyReachableRanges.subRangeMap(range)
-            for (entry in alreadyReachableMap.asMapOfRanges()) {
-                if (entry.value.apply(entry.key.upperEndpoint()) > parameters.baseNodeCost)
-                    return false
-            }
+            Range.closedOpen(
+                timeData.earliestReachableTime,
+                timeData.earliestReachableTime + timeData.maxDepartureDelayingWithoutConflict
+            )
+        val subMap = visitedRanges.subRangeMap(visitingRange)
+
+        // Check that the range is fully covered by the map
+        val uncovered = TreeRangeSet.create<Double>()
+        uncovered.add(visitingRange)
+        for (entry in subMap.asMapOfRanges()) uncovered.remove(entry.key)
+        if (!uncovered.isEmpty) return false
+
+        // Check that every range is considered "visited" with the given parameters
+        for (entry in subMap.asMapOfRanges()) {
```

Comment:
You could probably add this check in the previous loop in a boolean, and do:
`if (!uncovered.isEmpty || !isVisited) return false`
If it muddles the code too much, keeping it as is is fine as well.

Reply:
Fixed in 0c57be6390f29f51b60b1078138c60211d4dfd1f

There's also an extra corner case identified with the new unit test handled here